### ISSUE TITLE
Performance and profiling discussion

### ISSUE THREAD

**Issue by gmalette** [gadget-inc/wds#25]
## TL; DR: 

- Booting `esbuild-dev` currently costs ~130 ms only for itself.
- Requiring `validate-boot.ts` and not executing its code takes ~1800ms.
	- File System access from the child process appears to be negligible.
	- `esbuild` takes ~255 ms.
	- `module._compile` takes ~1100 ms.
	- There appears to be a ~370 ms overhead in `SyncWorker`.
	- There appears to be high variability in how long it takes to schedule a reload once a file changes.

Suggestions:
- Invest in reducing the number of files to `_compile`.
- Look into bootsnap-like caching of compiled files/bytecode if possible.
- Understand if there's actual overhead in `SyncWorker`, and why.
- Investigate reload scheduling variability.

# Understanding `esbuild-dev`'s performance

## Part 1: Big Blocks

At the moment I'm pretty clueless about the performance of `esbuild-dev`, how long I should expect anything to take, and where the time is spent.

To start, I'm interested in the following parts:  
- Boot of NodeJS, doing nothing. This is a fixed cost which we cannot avoid paying.  
- Boot of esbuild-dev. This could be optimized, if needs be.  
- Time between save and restart under `--watch --supervise` (small and large).
- Time for restart after `package.json` / `yarn.lock` changes.

These tests were measured on a:

MacBook Pro 14, 2021
32 GB Ram
10 Cores (8 performance/2 efficiency)

### Vanilla Node boot
To test this, I launched node processes with a file that only logs, and measure the time taken. This gives us the expected time required to boot a node process.

```
node src/noop.js #{run_number}
```

p50 = 30ms
p99 = 37ms

![image](https://user-images.githubusercontent.com/351915/145875818-8de3ca46-f30d-41bc-996f-1759971540c2.png)

### esbuild-dev boot
Similar to the previous exercise, but running `esbuild-dev`:

```
esbuild-dev src/noop.js #{run_number}
```

p50 = 138ms
p99 = 160ms

We do expect `esbuild-dev` to startup a file more slowly than `node` because there's two `node` process being started: one for `esbuild-dev` and one executed by it for the desired process (`src/noop.js`), and so our expectations are _at least_ 60ms, but we can see it's a bit slower than that here.

![image](https://user-images.githubusercontent.com/351915/145875856-60a652ff-b9b0-4f13-8063-0453393fc6bd.png)

### esbuild-dev supervise a small file
This test measures how long `esbuild-dev` takes to react and restart a process when changing a single small file.

`esbuild-dev --supervise src/noop.js #{run_number}`

In this test we measure two metrics: 
1. Delay between `touch` and the time `esbuild-dev` schedules the file to be reloaded
2. The time it takes between the scheduling and for the file to have been fully reloaded

The delay at #1 (between `touch` and scheduling) is susceptible to high variability, and I'm not sure what causes it. It's possible that the `fs.utimesSync` is not flushed right away.

**Schedule**
p50 = 114ms
p99 = 117ms

**Reload**
p50 = 117ms
p99 = 127ms

In this test we can see that the Reload time is very much in line with the `esbuild-dev` boot minus booting the initial `node` process.

![image](https://user-images.githubusercontent.com/351915/145875876-b68e19d7-38b4-493b-8f1f-f248cb0da32b.png)

### esbuild-dev supervise a larger codebase (with cache)
This test is is similar to the previous one but uses a larger codebase, in this case a modified version of Gadget's `api/tools/validate-boot.ts` file, to avoid running any code.

In this first test we measured with full use of the OS's file cache.

**Schedule**
p50 =117 ms
p99 = 124 ms

**Reload**
p50 = 699 ms
p99 = 799 ms

![image](https://user-images.githubusercontent.com/351915/145875881-8732d0ae-52e9-4d6a-971d-d4dbfe5f2114.png)

## esbuild-dev supervise a larger codebase (with cache)
This test is is similar to the previous one but uses a larger codebase, in this case Gadget's `api/tools/validate-boot.ts` file.

**This is probably the most common scenario**.

In this first test we measured with full use of the OS's file cache.

**Schedule**
p50 =117 ms
p99 = 124 ms

**Reload**
p50 = 699 ms
p99 = 799 ms

### esbuild-dev supervise a larger codebase (without cache)
This test is is similar to the previous one purges the file cache between each run. This scenario isn't excepted to be common, but it's still interesting.

Interestingly, we can see that the cache is resposible for a speedup of ~32%.

**Schedule**
p50 = 113 ms
p99 = 117 ms

**Reload**
p50 = 1,018 ms
p99 = 1,168 ms

![image](https://user-images.githubusercontent.com/351915/145875896-541ec7b7-fce4-4847-ab84-e641213dcdce.png)

## Part 2: Where Time is Spent

![image](https://user-images.githubusercontent.com/351915/145875923-69a2b5b5-a1fd-4e2c-8c52-31cf4b0162ca.png)

At first glance, it looks like `module._compile` is taking a long time. 

In fact, using the same modified `validate-boot.ts` file, which takes approx 1800 ms to load it, approx 1100 ms are attributable to `module._compile`. Of the remainder, 255 ms are attributable to `esbuild.build`. File system access is negligible (1.7ms). 

![image](https://user-images.githubusercontent.com/351915/145875932-1d1ec96d-94c9-40d5-80af-06a2089bbf6f.png)

From what I've been able to measure*, 369 ms were spent as `SyncWorker` communication overhead/delay.

* I have low relative confidence because of how log messages are passed from the worker to the main thread, apparently?
